### PR TITLE
Feature/json error factory

### LINKED ISSUE #10: Define and implement JSON-error-object factory
Define and implement a generic factory for an json error object. 

- https://jsonapi.org/format/#error-objects
- https://jsonapi.org/examples/#error-objects

### PR DESCRIPTION
**Pull request by Witti**

### CODE REVIEW

File: src/Factories/JsonErrorFactory.php [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+     * @inheritDoc
+     */
+    public static function createFromThrowable(Throwable $throwable, $status = 500): JsonError
+    {
+        $jsonError = new JsonError();
+        $jsonError->status = $status;
+        $jsonError->code = $throwable->getCode();
+        $jsonError->title = $throwable->getMessage();
+        $jsonError->detail = $throwable->getTraceAsString();
+        $jsonError->links = [
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: The Docs say that links should only contain `about`: https://jsonapi.org/format/#error-objects

File: src/Factories/JsonErrorFactory.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            'file' => $throwable->getFile(),
+            'line' => $throwable->getLine()
+        ];
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: The `id`, `source` and `meta` from the JsonError are never set.

File: src/Models/JsonError.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+class JsonError
+{
+    public string $id;
+    public string $code;
+    public int $status;
+    public string $title;
+    public string $detail;
+    public array $links;
+    public array $source;
+    public $meta;
```
Comment (CathrinAuer) [inline src/Models/JsonError.php]: Why has the `meta` no type? 🤔

File: src/Factories/JsonErrorFactory.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+        return $jsonError;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function createFromThrowables(array $throwables, $status = 500): array
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: I think that's problematic, because every `JsonError` would then have the same status code.

Comment (CathrinAuer) [reply src/Factories/JsonErrorFactory.php]: Solution: We create an own Exception class, where `id`, `source`, `meta` can also be set, so we are more flexible.
If our own Exception is not used, then we use `500` as a status code.

File: src/Factories/JsonErrorFactory.php [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+/**
+ * Class JsonErrorFactory
+ * @package WtfPhp\JsonApiErrors\Factories
+ */
+class JsonErrorFactory implements JsonErrorFactoryInterface
+{
+    /**
+     * @inheritDoc
+     */
+    public static function createFromThrowable(Throwable $throwable, $status = 500): JsonError
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: Why is there no type for `$status`?

File: src/Exceptions/JsonErrorException.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+     * @param string $message
+     * @param int $code
+     * @param Throwable|null $previous
+     * @param string $id
+     * @param int $statusCode
+     * @param array $meta
+     */
+    public function __construct(
+        $message = "",
+        $code = 0,
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Can you add the types for `$message` and `$code`?

File: src/Factories/JsonErrorFactory.php [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+     * @inheritDoc
+     */
+    public static function createFromThrowable(Throwable $throwable, int $status = 500): JsonError
+    {
+        $jsonError = new JsonError();
+        $jsonError->code = $throwable->getCode();
+        $jsonError->title = $throwable->getMessage();
+        $jsonError->detail = $throwable->getTraceAsString();
+        $jsonError->links = [
+            'about' => sprintf('%s at lint %s', $throwable->getFile(), $throwable->getLine())
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: We're not so happy about the link-section. It's supposed to be an actual link, the information about file and line should not go here.

File: src/Factories/JsonErrorFactory.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+/**
+ * Class JsonErrorFactory
+ * @package WtfPhp\JsonApiErrors\Factories
+ */
+class JsonErrorFactory implements JsonErrorFactoryInterface
+{
+    /**
+     * @inheritDoc
+     */
+    public static function createFromThrowable(Throwable $throwable, int $status = 500): JsonError
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: We don't need the `status`-param anymore 🙂

File: src/Factories/JsonErrorFactory.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+            $jsonError->meta = $throwable->getMeta() ?? [];
+        }
+
+        return $jsonError;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public static function createFromThrowables(array $throwables, int $status = 500): array
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: We don't need the status-param anymore 🙂

File: src/Exceptions/JsonErrorException.php [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        $this->id = $id;
+        $this->code = $code;
+        $this->meta = $meta;
+        $this->statusCode = $statusCode;
+    }
+
+    /**
+     * @return string
+     */
+    public function getId()
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Add return type

File: src/Exceptions/JsonErrorException.php [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return $this->id;
+    }
+
+    /**
+     * @return int
+     */
+    public function getStatusCode()
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Add return type

File: src/Exceptions/JsonErrorException.php [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        return $this->statusCode;
+    }
+
+    /**
+     * @return array
+     */
+    public function getMeta()
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Add return type

File: src/Exceptions/JsonErrorException.php [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * Class JsonErrorException
+ * @package WtfPhp\JsonApiErrors\Exceptions
+ */
+class JsonErrorException extends Exception
+{
+    protected string $id;
+    protected int $statusCode;
+    protected array $meta;
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Why not initialize the props already here with default values?

File: src/Exceptions/JsonErrorException.php [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * JsonErrorException constructor.
+     * @param string $message
+     * @param int $code
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Specification says the code should be a string 🙂

File: src/Exceptions/JsonErrorException.php [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+     * @param Throwable|null $previous
+     * @param string $id
+     * @param int $statusCode
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: Specification says the code should be a string 🙂

File: src/Factories/JsonErrorFactory.php (hunk 7 above)
Comment (CathrinAuer) [reply src/Factories/JsonErrorFactory.php]: We already have the information through the trace, isn't that enough?

File: src/Factories/JsonErrorFactory.php [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+        $jsonError->title = $throwable->getMessage();
+        $jsonError->detail = $throwable->getTraceAsString();
+        $jsonError->links = [
+            'about' => sprintf('%s at lint %s', $throwable->getFile(), $throwable->getLine())
+        ];
+
+        if ($throwable instanceof JsonErrorException) {
+            $jsonError->id = $throwable->getId() ?? '';
+            $jsonError->status = $throwable->getStatusCode() ?? $status;
+            $jsonError->meta = $throwable->getMeta() ?? [];
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: If we initialize the props we don't need the null-check.

File: src/Models/JsonError.php [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * Class JsonError
+ * @package WtfPhp\JsonApiErrors\Models
+ */
+class JsonError
+{
+    public string $id;
+    public string $code;
+    public int $status;
```
Comment (CathrinAuer) [inline src/Models/JsonError.php]: Should be string

File: src/Exceptions/JsonErrorException.php [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+     * @param array $meta
+     */
+    public function __construct(
+        $message = "",
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: single quotes

File: src/Exceptions/JsonErrorException.php [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+namespace WtfPhp\JsonApiErrors\Exceptions;
+
+use Exception;
+use Throwable;
+
+/**
+ * Class JsonErrorException
+ * @package WtfPhp\JsonApiErrors\Exceptions
+ */
+class JsonErrorException extends Exception
```
Comment (CathrinAuer) [inline src/Exceptions/JsonErrorException.php]: JsonApiErrorException? 🤔

File: src/Factories/JsonErrorFactory.php [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+
+use Throwable;
+use WtfPhp\JsonApiErrors\Exceptions\JsonErrorException;
+use WtfPhp\JsonApiErrors\Models\JsonError;
+
+/**
+ * Class JsonErrorFactory
+ * @package WtfPhp\JsonApiErrors\Factories
+ */
+class JsonErrorFactory implements JsonErrorFactoryInterface
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactory.php]: JsonApiErrorFactory? 🤔

File: src/Factories/JsonErrorFactoryInterface.php [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+namespace WtfPhp\JsonApiErrors\Factories;
+
+use Throwable;
+use WtfPhp\JsonApiErrors\Models\JsonError;
+
+/**
+ * Interface JsonErrorFactoryInterface
+ * @package WtfPhp\JsonApiErrors\Factories
+ */
+interface JsonErrorFactoryInterface
```
Comment (CathrinAuer) [inline src/Factories/JsonErrorFactoryInterface.php]: JsonApiErrorFactoryInterface? 🤔

File: src/Models/JsonError.php [hunk 22]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace WtfPhp\JsonApiErrors\Models;
+
+/**
+ * Class JsonError
+ * @package WtfPhp\JsonApiErrors\Models
+ */
+class JsonError
```
Comment (CathrinAuer) [inline src/Models/JsonError.php]: JsonApiError? 🤔

File: tests/JsonErrorFactoryTest.php [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+use Exception;
+use PHPUnit\Framework\TestCase;
+use WtfPhp\JsonApiErrors\Exceptions\JsonErrorException;
+use WtfPhp\JsonApiErrors\Factories\JsonErrorFactory;
+use WtfPhp\JsonApiErrors\Models\JsonError;
+
+class JsonErrorFactoryTest extends TestCase
+{
+    /** @test */
+    public function itShouldReturnAJsonErrorObject() {
```
Comment (CathrinAuer) [inline tests/JsonErrorFactoryTest.php]: All opening braces should go into the next line 🙂

File: tests/JsonErrorFactoryTest.php [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+        $e = new Exception('foo', 200);
+        $this->assertInstanceOf(JsonError::class, JsonErrorFactory::createFromThrowable($e));
```
Comment (CathrinAuer) [inline tests/JsonErrorFactoryTest.php]: Why not also check here if the `message` and the `code` have been set correctly?

File: tests/JsonErrorFactoryTest.php (hunk 23 above)
Comment (CathrinAuer) [reply tests/JsonErrorFactoryTest.php]: `itShouldReturnAJsonErrorObjectFromException`?

File: examples/FactoryExample.php [hunk 25]
```diff
@@ -0,0 +1,24 @@
+<?php
+
+use WtfPhp\JsonApiErrors\Exceptions\JsonErrorException;
+use WtfPhp\JsonApiErrors\Factories\JsonErrorFactory;
+use WtfPhp\JsonApiErrors\ThrowablesBag;
+
+require '../vendor/autoload.php';
+
+$bag = new ThrowablesBag();
+$exception = new Exception('foobar',123);
```
Comment (CathrinAuer) [inline examples/FactoryExample.php]: space missing after the comma

File: src/Factories/JsonErrorFactory.php (hunk 7 above)
Comment (Witti) [reply src/Factories/JsonErrorFactory.php]: I moved the about link to the JsonErrorException.

File: tests/JsonErrorFactoryTest.php (hunk 24 above)
Comment (Witti) [reply tests/JsonErrorFactoryTest.php]: I already implemented 2 tests for message and code.
`itShouldContainTheErrorCode` and `itShouldContainTheErrorTitle`

File: src/Models/JsonApiError.php [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+
+/**
+ * Class JsonApiError
+ * @package WtfPhp\JsonApiErrors\Models
+ */
+class JsonApiError
+{
+    public string $id;
+    public string $code;
+    public int $status;
```
Comment (CathrinAuer) [inline src/Models/JsonApiError.php]: string please 🙂

File: tests/JsonErrorFactoryTest.php [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace WtfPhp\JsonApiErrors\Tests;
+
+use Exception;
+use PHPUnit\Framework\TestCase;
+use WtfPhp\JsonApiErrors\Exceptions\JsonApiErrorException;
+use WtfPhp\JsonApiErrors\Factories\JsonApiErrorFactory;
+use WtfPhp\JsonApiErrors\Models\JsonApiError;
+
+class JsonErrorFactoryTest extends TestCase
```
Comment (CathrinAuer) [inline tests/JsonErrorFactoryTest.php]: `JsonApiErrorFactoryTest`

File: tests/JsonErrorFactoryTest.php [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+{
+    /** @test */
+    public function itShouldReturnAJsonErrorObjectFromException()
```
Comment (CathrinAuer) [inline tests/JsonErrorFactoryTest.php]: itShouldReturnA**JsonApiError**ObjectFromException

Could you also change the naming of the other functions?

File: src/Factories/JsonApiErrorFactory.php [hunk 29]
```diff
@@ -0,0 +1,50 @@
...
+ */
+class JsonApiErrorFactory implements JsonApiErrorFactoryInterface
+{
+    /**
+     * @inheritDoc
+     */
+    public static function createFromThrowable(Throwable $throwable): JsonApiError
+    {
+        $jsonError = new JsonApiError();
+        $jsonError->code = $throwable->getCode();
```
Comment (CathrinAuer) [inline src/Factories/JsonApiErrorFactory.php]: shouldn't we cast the code to a string here? The json:api specification says "an application-specific error code, expressed as a string value" 🙂

**Review by CathrinAuer — APPROVED** [review]

**Review by DanielaHuber — APPROVED** [review]

**Pull request merged**